cranelift-codegen/src/simple_preopt.rs

```diff
@@ -536,11 +542,245 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
+            destination,
+            ref args,
+        } => {
+            let next_ebb = if let Some(next_ebb) = pos.func.layout.next_ebb(ebb) {
+                next_ebb
+            } else {
+                return;
+            };
+
+            if destination == next_ebb {
+                return;
+            }
+
+            let prev_inst = if let Some(prev_inst) = pos.func.layout.prev_inst(inst) {
+                prev_inst
+            } else {
+                return;
+            };
+
+            let prev_inst_data = &pos.func.dfg[prev_inst];
+
+            if let Some(prev_dest) = prev_inst_data.branch_destination() {
+                if prev_dest != next_ebb {
+                    return;
+                }
+            } else {
+                return;
+            }
+
+            match prev_inst_data {
+                InstructionData::Branch {
+                    opcode,
+                    args: ref prev_args,
+                    destination: cond_dest,
+                } => {
+                    let cond_arg = {
+                        let args = pos.func.dfg.inst_args(prev_inst);
+                        args[0]
+                    };
+
+                    let kind = match opcode {
+                        Opcode::Brz => BranchOrderKind::BrzToBrnz(cond_arg),
+                        Opcode::Brnz => BranchOrderKind::BrnzToBrz(cond_arg),
+                        _ => panic!("unexpected opcode"),
+                    };
+
+                    BranchOrderInfo {
+                        term_inst: inst,
+                        term_inst_args: args.clone(),
+                        term_dest: destination,
+                        cond_inst: prev_inst,
+                        cond_inst_args: prev_args.clone(),
+                        cond_dest: *cond_dest,
+                        kind: kind,
+                    }
+                }
+                InstructionData::BranchIcmp {
+                    opcode: Opcode::BrIcmp,
+                    cond,
+                    destination: cond_dest,
+                    args: ref prev_args,
+                } => {
+                    let (x_arg, y_arg) = {
+                        let args = pos.func.dfg.inst_args(prev_inst);
+                        (args[0], args[1])
+                    };
+                    BranchOrderInfo {
+                        term_inst: inst,
+                        term_inst_args: args.clone(),
+                        term_dest: destination,
+                        cond_inst: prev_inst,
+                        cond_inst_args: prev_args.clone(),
+                        cond_dest: *cond_dest,
+                        kind: BranchOrderKind::InvertIcmpCond(*cond, x_arg, y_arg),
+                    }
+                }
+                _ => return,
+            }
+        }
+
+        _ => return,
+    };
+
+    let cond_args = {
+        info.cond_inst_args
+            .as_slice(&pos.func.dfg.value_lists)
+            .to_vec()
```

Comment:
It seems all the uses of `cond_args` can use a slice; could we remove the vec heap allocation here?

Reply:
The trouble is that just below this I borrow `pos.func.dfg` as mutable. I haven't been able to figure out how to appease the borrow checker in this case. I'm all ears if you have suggestions though.

Reply:
Sounds fair, I asked just in case.